### PR TITLE
Refactored value objects

### PR DESCRIPTION
**Pull request by petrknap**
https://redmine.havelholding.cz/issues/1951

### CODE REVIEW

File: src/ValueObject/CountryCode.php [hunk 1]
```diff
@@ -1,37 +1,20 @@
...
+    protected static function filter($value)
     {
-        if ( ! is_string($value) || ! preg_match('/[A-Z]{2}/', $value)) {
-            throw new \InvalidArgumentException('Value `' . $value . '` is not valid ISO 3166-1 (alpha-2) country code');
-        }
-
-        $this->value = $value;
-    }
+        $nonEmptyString = parent::filter($value);
+        $countryCodeCandidate = strtoupper(trim($nonEmptyString));
```
Comment (petrknap) [inline src/ValueObject/CountryCode.php]: Is it okay to apply `strtoupper`?

File: src/ValueObject/CurrencyCode.php [hunk 2]
```diff
@@ -1,43 +1,20 @@
...
-     * CurrencyCode constructor.
-     *
-     * @param string $value
-     */
-    public function __construct($value)
+    protected static function filter($value)
     {
-        $value = trim($value);
+        $nonEmptyString = parent::filter($value);
+        $currencyCodeCandidate = strtoupper(trim($nonEmptyString));
```
Comment (petrknap) [inline src/ValueObject/CurrencyCode.php]: Is it okay to apply `strtoupper`?

File: src/ValueObject/EnumValueObject.php [hunk 3]
```diff
@@ -1,45 +1,27 @@
...
     /**
-     * @param string $value
+     * @return string[]
      */
-    public function __construct($value)
-    {
-        if ( ! in_array($value, static::getOptions(), true)) {
-            throw new \InvalidArgumentException(sprintf('%s in not valid value', $value));
-        }
+    abstract public static function cases(): array;
```
Comment (petrknap) [inline src/ValueObject/EnumValueObject.php:15]: Renamed to met <https://www.php.net/manual/en/unitenum.cases.php>

File: src/ValueObject/LanguageCode.php [hunk 4]
```diff
@@ -1,44 +1,20 @@
...
-     * CurrencyCode constructor.
-     *
-     * @param string $value
-     */
-    public function __construct($value)
+    protected static function filter($value)
     {
-        $value = trim($value);
+        $nonEmptyString = parent::filter($value);
+        $languageCodeCandidate = strtolower(trim($nonEmptyString));
```
Comment (petrknap) [inline src/ValueObject/LanguageCode.php]: Is it okay to apply `strtolower`?

File: src/ValueObject/SecureUrl.php [hunk 5]
```diff
@@ -1,43 +1,19 @@
...
+    protected static function filter($value)
     {
-        if ( ! filter_var($url, FILTER_VALIDATE_URL)) {
-            throw new \InvalidArgumentException('Url is in incorrect format');
-        }
+        $url = parent::filter($value);
+
         if (strpos($url, 'https://') !== 0) {
-            throw new \InvalidArgumentException('Url must start with "https://"');
+            throw self::invalidValue('"https://" prefixed', $url);
```
Comment (petrknap) [inline src/ValueObject/SecureUrl.php]: Any idea for better message? `"https://" prefixed URL`?

File: src/ValueObject/CountryCode.php (hunk 1 above)
Comment (Triplkrypl) [reply src/ValueObject/CountryCode.php]: We can remove conversion and validate input string only for upper case characters.
With validation we can prevent language substitution by mistake, example 'sk' language is not 'SK' country, this is reason why ISO language is lower case and ISO countries is upper case.

File: src/ValueObject/SecureUrl.php (hunk 5 above)
Comment (Triplkrypl) [reply src/ValueObject/SecureUrl.php]: One example for inspiration: https://github.com/simple-as-fuck/php-validator/blob/44f2b59996e52d571c5534476a4cdba6760fa6e6/src/Rule/Url/UrlRule.php#L79.
I not have better idea.

File: src/ValueObject/BaseValueObject.php [hunk 6]
```diff
@@ -57,18 +59,13 @@ public function getValue()
...
      *
      * @return TValue
      *
      * @throws InvalidArgumentException
      */
-    protected static function filter($value)
-    {
-        throw self::invalidValue('expected');
-    }
+    abstract protected static function filter($value);
```
Comment (joseftraxler) [inline src/ValueObject/BaseValueObject.php:68]: Why does it name filter? I'd like to rename it to `adjustValue` or `validateAndAdjustValue` or something like this.
```suggestion
    abstract protected static function validateAndAdjustValue($value);
```

File: src/ValueObject/CurrencyCode.php (hunk 2 above)
Comment (joseftraxler) [reply src/ValueObject/CurrencyCode.php]: Don't know why not. IMHO it's OK.

File: src/ValueObject/BaseValueObject.php [hunk 7]
```diff
@@ -57,18 +59,13 @@ public function getValue()
     }
 
     /**
-     * @note should be abstract
-     *
      * @param TValue|mixed $value
```
Comment (joseftraxler) [inline src/ValueObject/BaseValueObject.php]:
```suggestion
     * @param mixed $value
```
mixed is enough

**Review by joseftraxler — CHANGES_REQUESTED** [review]
I just wanna rename filter method to something better. All changes have tests, well done 👍.

File: src/ValueObject/BaseValueObject.php (hunk 6 above)
Comment (petrknap) [reply src/ValueObject/BaseValueObject.php:68]: Method `filter` refers to <https://www.php.net/manual/en/function.filter-var.php> - in ideal world the implementation should be like:

```php
protected static function filter(mixed $value): int
{
    $intCandidate = filter_var($value, FILTER_VALIDATE_INT);
    if ($intCandidate === false) {
        throw self::invalidValue('int');
    }
    return $intCandidate;
}
```

Comment (joseftraxler) [reply src/ValueObject/BaseValueObject.php:68]: OK, i don't wanna waste my time, so keep it as you wish.

**Review by joseftraxler — APPROVED** [review]

**Review by Triplkrypl — APPROVED** [review]

**Pull request merged**